server/broadcast.go

```diff
@@ -347,13 +359,30 @@ func transcodeSegment(cxn *rtmpConnection, seg *stream.HLSSegment, name string,
 	// send segment to the orchestrator
 	glog.V(common.DEBUG).Infof("Submitting segment nonce=%d manifestID=%s seqNo=%d orch=%s", nonce, cxn.mid, seg.SeqNo, sess.OrchestratorInfo.Transcoder)
 
-	res, err := SubmitSegment(sess, seg, nonce)
-	if err != nil || res == nil {
+	res, tErr := SubmitSegment(sess, seg, nonce)
+	// If TicketParams have expired, refresh them
+	if tErr != nil && tErr == pm.ErrTicketParamsExpired {
+		glog.V(common.VERBOSE).Infof("ticketparams expired, refreshing for orchestrator=%v", sess.OrchestratorInfo.Transcoder)
+		oURL, err := url.Parse(sess.OrchestratorInfo.Transcoder)
+		if err != nil {
+			cxn.sessManager.removeSession(sess)
+			return nil, fmt.Errorf("invalid orchestrator URL url=%v", sess.OrchestratorInfo.Transcoder)
+		}
+		oInfo, err := getOrchestratorInfoRPC(context.Background(), sess.Broadcaster, oURL)
+		if err != nil {
+			cxn.sessManager.removeSession(sess)
+			return nil, fmt.Errorf("unable to refresh ticketparams for orchestrator=%v, err=%v", sess.OrchestratorInfo.Transcoder, err)
+		}
+		sess.OrchestratorInfo = oInfo
```

Comment:
Could we wrap all of this into a helper function `refreshSession()` that uses the existing `updateSession()` function to return a copied session with updated fields (if the `OrchestratorInfo` changes we might need to update additional fields in the session). Something like:

```
func refreshSession(sess *BroadcastSession) (*BroadcastSession, error) {
	uri, err := url.Parse(sess.OrchestratorInfo.Transcoder)
	if err != nil {
		return nil, err
	}

	refreshTimeout := 3 * time.Second
	ctx, cancel := context.WithTimeout(context.Background(), refreshTimeout)
	defer cancel()

	oInfo, err := GetOrchestratorInfo(ctx, sess.Broadcaster, uri)
	if err != nil {
		return nil, err
	}

        // Create dummy result
	res := &ReceivedTranscodeResult{
		LatencyScore: sess.LatencyScore,
		Info:         oInfo,
	}

	return updateSession(sess, res), nil
}
```

If we use a helper, we would only need to worry about calling `removeSession()` if an error is returned by the helper.

Reply:
fixed in c274f67c5296ea35f0c635d0b86ef5780506cf4a & ebb7532edd782df1b2352b1beb5e46159687474b